README.md
# numerical-calculus
Libraries for use in differentiation, integration and root finding. 
Can be used for basic calculus techniques in small intervals.

## Numerical Differentiation
Libraries for differentiation include:
- Central difference method
- Forward difference method
- Backward difference method
- Second derivative calculation

There is an included demo executable that when run will use the example of a sine function at PI/3

## Numerical Integration
Libraries for integration include:
- Trapezoidal method
- Simpsons Rule
- Newton-Cotes (Degree 3) Formulae

Another demo is included that demonstrates integrating a function log(1+x)/x over a small interval.
(Something to note that these formulae are only accurate over small intervals, especially the Trapezoidal method - as you will see from the demo).

## Root Finding
Libraries for root finding include:
- Newton-Raphson Method
- False Position

Included is root_demo_exec which demonstrates the accuracy of False Position in comparison to the Newton-Raphson Method of finding the roots of y(x) = x^2 - 5.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Pre-release]

## [1.0.0] - [1.1.0] : 2017-06-29
### Added
- Differentiation Libraries.
- Integration Libraries.
- Root Finding Libraries.
- Changelog.
- package.json file.
- Demos for Differentiation, Integration and Root Finding.
- Wiki Section.
### Changed
- Document structure.
- README.md information.


diff_demo.c
// demo of the differentiation libraries
// author: Cohen Robinson
// date: 27/06/18

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "differentiation.h"

int main() {
  //define file
  FILE *fOut;

  fOut = fopen("comparemethods.txt","w");

  int n = 4;
  int i;
  double x = PI/3;
  double h;
  function sin;

  //prints out titles
  fprintf(fOut,"%-7s %-7s %-7s %-7s\n","n","FDM","BDM","CDM");

  //loops through all values of n
  for(i=1;i<=n;i++){
  	h=pow(10,-i);
  	fprintf(fOut,"%-5d %-5.5f %-5.5f %-5.5f\n", i, forward_difference(sin, x, h), backward_difference(sin, x, h), central_difference(sin, x, h));
  }

  //Extention
  //sin x is the second derivative of -sinx
  fprintf(fOut,"\n%-6s %-6s %-6s\n","n","2NDFX","cos(x)");
  i=1;
  for ( i = 1; i <= n; i++){
  	h = pow(10,-i);
  	fprintf(fOut,"%-5d %-5.5f %5.5f\n", i, second_derivative(sin, x, h), -sin(x));
  }
	//will converge to the correct value as n increases

  return 0;
}


root_demo.c
// demo for root_finding libraries
// author: Cohen Robinson
// date: 27/06/18

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "root_finding.h"

// function to be demonstrated
double xsqminfive( double x){
  return x * x - 5;
}

int main(){
  double xacc = 0.1;  // required precision
  double trueroot = sqrt(5);  // actual value for root
  double nr_root = newton_raph(xsqminfive, 0, 3, xacc);
  double fp_root = false_pos(xsqminfive, 0, 3, 0.001);

  printf("Function: y(x) = x^2 - 5\n");
  printf("Expected root: x = sqrt(5)\n"); // negative root outside of 0 - 3 bound
  printf("Newton-Raphson: %f\n", nr_root);
  printf("False-Position: %f\n", fp_root);
  printf("Error in Newton-Raphson: %f\n", fabs(trueroot-nr_root));
  printf("Error in False Position: %f\n", fabs(trueroot-fp_root));

  return 0;
}


integration.h
// libraries for various numerical integration methods
// author: Cohen Robinson
// date: 27/06/18

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define PI 3.1415926535897932384

typedef double (function)(double);

// defintion for trapizoidal integration
// assumes linear across region
double trapezoidal(double left_bound, double right_bound, function f, int iterations)
{
  // define variables
	double x = left_bound;     // start at left_bound
	double h = (right_bound - left_bound) / iterations;    // step_size

	// sum over iterations
	double sum = f(0) + f(iterations);
	for (int i = 1; i < iterations; i++){
		sum += 2 * f(x);
    x += h;
	}

	return h * sum / 2;
}

// simpsons rule for integration assumes that f is a polynomial
double simpson(double left_bound, double right_bound, function f, int iterations)
{
  double x = left_bound;     // start at left_bound
	double h = (right_bound - left_bound) / iterations;    // step_size

  // sum over iteractions
  double sum = f(0) + f(iterations);    // m of 1
  int m = 4;    // to add constant to terms
  for (int i = 1; i < iterations; i++){
    sum += m * f(x);
    if (m == 4){
      m = 2;
    }
    else{
      m = 4;
    }
    x += h;
  }

  return h * sum / 3;
}

// newton cotes formulae assumes
// also to degree of 3
double newtoncotes(double left_bound, double right_bound, function f, int iterations)
{
	if (iterations <= 2){ // This insures that there is a minimum of 3 partitions.
		printf("ERROR: Please choose iterations greater than 2!\n");
    exit(0);
	}
	else {
		if (iterations % 3 != 0){
			printf("ERROR: Please choose iterations divisible by 3!\n");
      exit(0);
		}
	}

	unsigned int k = iterations / 3;

	double x = left_bound;
	double h = (right_bound - left_bound) / iterations;
	double current_sum = 0.0;
	int i;
	for (i = 0; i < k; i++){
		current_sum += 3 * h * (f( x + (3 * i)*h ) + 3.0 * f( x + (3 * i + 1)*h ) + 3.0 * f( x + (3 * i + 2)* h) + f( x + (3 * i + 3)*h)) / 8.0;
	}

	return current_sum; // All partitions have been summed.
}


differentiation.h
// libraries for use with numerical differentiation
// author: Cohen Robinson
// date: 27/06/18

#include <stdlib.h>
#include <stdio.h>
#include <math.h>

#define PI 3.14159265

// define function
typedef double (function)(double);

// define central_difference method
double central_difference(function f, double x, double h)
{
	return (f(x + h) - f(x - h)) / (2 * h);
}

// define backward difference method
double backward_difference(function f, double x, double h)
{
  return (f(x) - f(x - h)) / h;
}

// define forward difference method
double forward_difference(function f, double x, double h)
{
  return (f(x + h) - f(x)) / h;
}

// define factorial function
double factorial(int n)
{
  double fac = 1.;
  for (int i = 1; i <= n; i++){
    fac *= i;
  }
  return fac;
}

//second derivative - EXTENSION
double second_derivative(function f, double x, double h)
{
	return (f(x + h) + f(x - h) - 2 * f(x))/(pow(h,2));
}


library/root_finding.h
// libraries for use in root finding
// author: Cohen Robinson
// date: 27/06/18

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double grad( double (*function)(double), double x){
  double sep = 0.00001; // delta x
  double y2 = function(x+sep);
  double y1 = function(x-sep);
  double grad = (y2-y1)/(2*sep);
  return grad;
}

// function that runs the newton_raphison method of root finding
double newton_raph( double (*function)(double), double x1, double x2, double xacc){

  double xold = x1;   // begin at x1 and move to x2
  double xnew = xold - function(xold)/grad(function, xold);  // calc first root
  int counter = 1;

  while (fabs(xnew-xold) >= fabs(xacc)){

    xold = xnew;
    xnew = xold - function(xold)/grad(function, xold);
    counter++;
  }
  if (xnew >= x1 && xnew <= x2){
    printf("The number of iterations in the Newton-Raphson method is: %d\n", counter);
    return xnew;
  }
  else{
    if (x1 < x2){
      return newton_raph( function, x1+0.01, x2, xacc);
    }
    else{
      printf("Error in newton_raph: Failed to find a root in the given range!\n");
      return 0;
    }
  }
}

// function that runs the false position method
double false_pos( double(*function)(double), double xl, double xr, double yacc){
  int counter = 0;
  double x_new = xr;
  while (fabs(function(x_new)) >= yacc){     // iterate until function converges within accuracy
    double y1 = function(xl);
    double y2 = function(xr);
    if (y1 * y2 > 0){
      printf("Error in false_pos: You have not chosen two points where a line between them crosses the x-axis.\n");
      return 0;
    }
    x_new = xr - y2 * (xr - xl) / (y2 - y1);
    if (y1 * function(x_new) > 0){
      xl = x_new;
    }
    else{
      xr = x_new;
    }
    counter++;
  }
  printf("The number of iterations in False Position method were: %d\n", counter);
  return x_new;
}


examples/integration_demo.c
// demonstration of integration libraries
// author: Cohen Robinson
// date: 27/06/18

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "integration.h"

double f1(double x);

int main()
{
  FILE *fOut;

	fOut=fopen("num_integration.txt","w");

  double analyticResult = PI * PI / 12.0;

  int N;
	int i;
	double x_left = 1 * pow(10, -20);
	double x_right = 1;
  function f1;
	//print to file
	fprintf(fOut, "%-5s %-15s %-15s %-15s\n","Loop","Trapezoidal","Simpson","Newton-Cotes");
	//print to file
	for(i = 1; i < 12 ; i++) {
		//define n_partitions
		N = 3 * i;
		//print to file
		fprintf(fOut,"%-5d %-.13f %-.13f %.13f\n", i+1, analyticResult-trapezoidal(x_left, x_right, f1, N),analyticResult-simpson(x_left, x_right, f1, N),analyticResult-newtoncotes(x_left, x_right, f1, N));
		//print each error to the terminal
		printf("\nError in Trapezoidal in %d loop: %f",i+1,analyticResult-trapezoidal(x_left, x_right, f1, N));

		printf("\nError in Simpson in %d loop: %f",i+1,analyticResult-simpson(x_left, x_right, f1, N));

		printf("\nError in Newton-Cotes (Degree 3) in %d loop: %f", i+1 , analyticResult-newtoncotes(x_left, x_right, f1, N));

	}

  getchar();
  return 0;
}

// test function
double f1(double x)
{
  if (x == 0.0){
    return 1.0;
  }
  return log(1 + x) / x;      // natural log
}
